File: packages/butler-agent/src/agent/tools/run-command/run_command/evidence.ts
import { extname } from "path";
import {
  createEvidenceCapabilityReceipt,
  validationCapabilityReceipt,
} from "../../../output/evidence-capability-ledger.ts";
import { evidenceReceipt } from "../../../tool-support/executor-support.ts";
import type { EvidenceReceipt, PublicWorkObligationKind } from "../../../turn/native-tool-types.ts";
import type { EvidenceCapabilityReceipt } from "../../../output/evidence-capability-ledger.ts";

type CommandExecutionStatus = "succeeded" | "failed" | "timed_out";
type CommandArtifactKind = "csv_file" | "table_file" | "chart_file" | "file";

export interface CommandArtifactEvidence {
  path: string;
  artifact_kind: CommandArtifactKind;
  size_bytes: number;
  modified_at: string;
}

export interface CommandValidationEvidence {
  suite: string;
  result: "passed" | "failed" | "partial" | "skipped";
  failure_summary?: string;
}

export function commandEvidenceReceipts(input: {
  success: boolean;
  artifacts: CommandArtifactEvidence[];
}): EvidenceReceipt[] {
  const receipts: EvidenceReceipt[] = [
    evidenceReceipt({
      producerName: "run_command",
      receiptType: "execution",
      summary: input.success
        ? "A local command executed successfully."
        : "A local command was executed but did not complete successfully.",
      covers: ["execution_result"],
      verified: input.success,
      satisfies: input.success ? ["command_executed"] : [],
    }),
  ];
  if (input.artifacts.length > 0) {
    const satisfies = new Set<PublicWorkObligationKind>(["durable_artifact"]);
    if (input.artifacts.some((artifact) =>
      artifact.artifact_kind === "csv_file" || artifact.artifact_kind === "table_file",
    )) {
      satisfies.add("data_table_created");
    }
    if (input.artifacts.some((artifact) => artifact.artifact_kind === "chart_file")) {
      satisfies.add("chart_rendered");
    }
    receipts.push(evidenceReceipt({
      producerName: "run_command",
      receiptType: "deliverable",
      summary: "The command produced verified durable output file evidence.",
      covers: ["durable_deliverable"],
      artifacts: input.artifacts.map((artifact) => ({
        label: artifact.path,
        path: artifact.path,
        mediaType: commandArtifactMediaType(artifact),
        role: commandArtifactRole(artifact),
      })),
      satisfies: [...satisfies],
      metrics: {
        artifact_count: input.artifacts.length,
      },
    }));
  }
  return receipts;
}

function commandExecutionStatus(input: {
  exitCode: number | null;
  timedOut: boolean;
}): CommandExecutionStatus {
  if (input.timedOut) return "timed_out";
  return input.exitCode === 0 ? "succeeded" : "failed";
}

function exitCodeLabel(exitCode: number | null): string {
  return exitCode === null ? "none" : String(exitCode);
}

function safeArtifactPath(path: string): string | null {
  const trimmed = path.trim();
  if (!trimmed || trimmed.startsWith("/") || trimmed.startsWith("~")) return null;
  if (/^[A-Za-z]:[\\/]/u.test(trimmed)) return null;
  if (trimmed.split(/[\\/]+/u).includes("..")) return null;
  return trimmed;
}

export function commandEvidenceCapabilityReceipts(input: {
  exitCode: number | null;
  timedOut: boolean;
  outputSuppressed: boolean;
  outputBudgeted: boolean;
  artifacts?: CommandArtifactEvidence[];
  validations?: CommandValidationEvidence[];
}): EvidenceCapabilityReceipt[] {
  const status = commandExecutionStatus(input);
  const verified = status === "succeeded";
  const receipts: EvidenceCapabilityReceipt[] = [
    createEvidenceCapabilityReceipt({
      producer: { kind: "tool", name: "run_command" },
      capability: "command_executed",
      evidence_kind: "execution_result",
      maturity: verified ? "verified" : status === "timed_out" ? "candidate" : "rejected",
      verified,
      confidence: verified ? 1 : status === "timed_out" ? 0.45 : 0.65,
      summary: `Command execution ${status} with exit code ${exitCodeLabel(input.exitCode)}.`,
      scope: {
        status,
        exit_code: input.exitCode,
        timed_out: input.timedOut,
        output_suppressed: input.outputSuppressed,
        output_budgeted: input.outputBudgeted,
      },
      ...(verified ? { satisfies: ["command_executed" as const] } : {}),
      limitations: verified ? [] : ["Command execution did not complete successfully."],
    }),
  ];

  for (const validation of input.validations ?? []) {
    receipts.push(validationCapabilityReceipt({
      producer: { kind: "tool", name: "run_command" },
      suite: validation.suite,
      result: validation.result,
      failureSummary: validation.failure_summary,
    }));
  }

  for (const artifact of input.artifacts ?? []) {
    receipts.push(...artifactCapabilityReceipts(artifact));
  }
  return receipts;
}

function artifactCapabilityReceipts(artifact: CommandArtifactEvidence): EvidenceCapabilityReceipt[] {
  const path = safeArtifactPath(artifact.path);
  if (!path) return [];
  const reference = { label: path, path };
  const scope = {
    artifact_kind: artifact.artifact_kind,
    size_bytes: artifact.size_bytes,
    modified_at: artifact.modified_at,
  };
  const receipts = [
    createEvidenceCapabilityReceipt({
      producer: { kind: "tool", name: "run_command" },
      capability: "durable_artifact",
      evidence_kind: "artifact",
      maturity: "verified",
      verified: true,
      confidence: 0.9,
      summary: "Command produced a verified durable artifact reference.",
      scope,
      references: [reference],
      satisfies: ["durable_artifact"],
    }),
  ];
  if (artifact.artifact_kind === "csv_file" || artifact.artifact_kind === "table_file") {
    receipts.push(createEvidenceCapabilityReceipt({
      producer: { kind: "tool", name: "run_command" },
      capability: "data_table_created",
      evidence_kind: "data_table",
      maturity: "verified",
      verified: true,
      confidence: 0.9,
      summary: "Command produced a verified structured table artifact.",
      scope,
      references: [reference],
      satisfies: ["data_table_created"],
    }));
  }
  if (artifact.artifact_kind === "chart_file") {
    receipts.push(createEvidenceCapabilityReceipt({
      producer: { kind: "tool", name: "run_command" },
      capability: "chart_rendered",
      evidence_kind: "chart",
      maturity: "verified",
      verified: true,
      confidence: 0.9,
      summary: "Command produced a verified chart artifact.",
      scope,
      references: [reference],
      satisfies: ["chart_rendered"],
    }));
  }
  return receipts;
}

function commandArtifactMediaType(artifact: CommandArtifactEvidence): string {
  const ext = extname(artifact.path).toLocaleLowerCase("en-US");
  if (artifact.artifact_kind === "csv_file") return "text/csv";
  if (artifact.artifact_kind === "table_file") return "text/tab-separated-values";
  if (ext === ".png") return "image/png";
  if (ext === ".jpg" || ext === ".jpeg") return "image/jpeg";
  if (ext === ".webp") return "image/webp";
  if (ext === ".svg") return "image/svg+xml";
  if (ext === ".pdf") return "application/pdf";
  return "application/octet-stream";
}

function commandArtifactRole(artifact: CommandArtifactEvidence): string {
  if (artifact.artifact_kind === "csv_file" || artifact.artifact_kind === "table_file") return "table";
  if (artifact.artifact_kind === "chart_file") return "chart";
  return "file";
}

File: packages/butler-agent/rust/agent/src/host/guided_command/evidence.rs
use chrono::{DateTime, SecondsFormat, Utc};
use serde_json::{Value, json};
use uuid::Uuid;

use super::artifacts::Artifact;

fn now() -> String {
    DateTime::<Utc>::from(std::time::SystemTime::now()).to_rfc3339_opts(SecondsFormat::Millis, true)
}
fn id(prefix: &str) -> String {
    format!("{}-{}", prefix, &Uuid::new_v4().to_string()[..12])
}

pub(super) fn receipts(success: bool, artifacts: &[Artifact]) -> Vec<Value> {
    let mut result = vec![json!({
        "schema":"butler.evidence-receipt.v1", "id":id("receipt"),
        "producer":{"kind":"tool","name":"run_command"},
        "receiptType":"execution", "verified":success, "covers":["execution_result"],
        "summary":if success { "A local command executed successfully." } else { "A local command was executed but did not complete successfully." },
        "references":[], "satisfies":if success { vec!["command_executed"] } else { Vec::new() },
    })];
    if !artifacts.is_empty() {
        let table = artifacts
            .iter()
            .any(|a| matches!(a.artifact_kind, "csv_file" | "table_file"));
        let chart = artifacts.iter().any(|a| a.artifact_kind == "chart_file");
        let mut satisfies = vec!["durable_artifact"];
        if table {
            satisfies.push("data_table_created");
        }
        if chart {
            satisfies.push("chart_rendered");
        }
        result.push(json!({
            "schema":"butler.evidence-receipt.v1", "id":id("receipt"),
            "producer":{"kind":"tool","name":"run_command"},
            "receiptType":"deliverable", "verified":true, "covers":["durable_deliverable"],
            "summary":"The command produced verified durable output file evidence.",
            "references":[], "satisfies":satisfies,
            "artifacts":artifacts.iter().map(|a| json!({
                "label":a.path,"path":a.path,"mediaType":media_type(a),"role":role(a)
            })).collect::<Vec<_>>(),
            "metrics":{"artifact_count":artifacts.len()}
        }));
    }
    result
}

pub(super) fn capability_receipts(
    exit: Option<i32>,
    timed_out: bool,
    suppressed: bool,
    budgeted: bool,
    artifacts: &[Artifact],
) -> Vec<Value> {
    let status = if timed_out {
        "timed_out"
    } else if exit == Some(0) {
        "succeeded"
    } else {
        "failed"
    };
    let verified = status == "succeeded";
    let label = exit.map_or_else(|| "none".into(), |exit| exit.to_string());
    let mut first = base(
        "command_executed",
        "execution_result",
        if verified {
            "verified"
        } else if timed_out {
            "candidate"
        } else {
            "rejected"
        },
        verified,
        if verified {
            1.0
        } else if timed_out {
            0.45
        } else {
            0.65
        },
        &format!("Command execution {status} with exit code {label}."),
    );
    first["scope"] = json!({"status":status,"exit_code":exit,"timed_out":timed_out,
        "output_suppressed":suppressed,"output_budgeted":budgeted});
    if verified {
        first["satisfies"] = json!(["command_executed"]);
    } else {
        first["limitations"] = json!(["Command execution did not complete successfully."]);
    }
    let mut result = vec![first];
    for artifact in artifacts {
        if !safe(&artifact.path) {
            continue;
        }
        let reference = json!([{"label":artifact.path,"path":artifact.path}]);
        let scope = json!({"artifact_kind":artifact.artifact_kind,"size_bytes":artifact.size_bytes,"modified_at":artifact.modified_at});
        let mut receipt = base(
            "durable_artifact",
            "artifact",
            "verified",
            true,
            0.9,
            "Command produced a verified durable artifact reference.",
        );
        receipt["scope"] = scope.clone();
        receipt["references"] = reference.clone();
        receipt["satisfies"] = json!(["durable_artifact"]);
        result.push(receipt);
        let extra = match artifact.artifact_kind {
            "csv_file" | "table_file" => Some((
                "data_table_created",
                "data_table",
                "Command produced a verified structured table artifact.",
            )),
            "chart_file" => Some((
                "chart_rendered",
                "chart",
                "Command produced a verified chart artifact.",
            )),
            _ => None,
        };
        if let Some((capability, kind, summary)) = extra {
            let mut receipt = base(capability, kind, "verified", true, 0.9, summary);
            receipt["scope"] = scope;
            receipt["references"] = reference;
            receipt["satisfies"] = json!([capability]);
            result.push(receipt);
        }
    }
    result
}

fn base(
    capability: &str,
    kind: &str,
    maturity: &str,
    verified: bool,
    confidence: f64,
    summary: &str,
) -> Value {
    json!({"receipt_id":id("ecr"),"schema_version":"evidence-capability.v1",
        "producer":{"kind":"tool","name":"run_command"},
        "capability":capability,"evidence_kind":kind,"maturity":maturity,
        "confidence":confidence,"verified":verified,"summary":summary,
        "references":[],"limitations":[],"created_at":now()})
}
fn safe(path: &str) -> bool {
    let path = crate::public_text::trim_js_whitespace(path);
    !path.is_empty()
        && !path.starts_with('/')
        && !path.starts_with('~')
        && path.as_bytes().get(1).is_none_or(|ch| *ch != b':')
        && !path.split(['/', '\\']).any(|part| part == "..")
}
fn media_type(artifact: &Artifact) -> &'static str {
    match artifact
        .path
        .rsplit('.')
        .next()
        .unwrap_or("")
        .to_ascii_lowercase()
        .as_str()
    {
        "csv" => "text/csv",
        "tsv" => "text/tab-separated-values",
        "png" => "image/png",
        "jpg" | "jpeg" => "image/jpeg",
        "webp" => "image/webp",
        "svg" => "image/svg+xml",
        "pdf" => "application/pdf",
        _ => "application/octet-stream",
    }
}
fn role(artifact: &Artifact) -> &'static str {
    match artifact.artifact_kind {
        "csv_file" | "table_file" => "table",
        "chart_file" => "chart",
        _ => "file",
    }
}
